sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1251 @@
+        X, y, labels = indexable(X, y, labels)
+        n = _num_samples(X)
+        if self.n_folds > n:
+            raise ValueError(
+                ("Cannot have number of folds n_folds={0} greater"
+                 " than the number of samples: {1}.").format(self.n_folds, n))
+
+        for train, test in super(_BaseKFold, self).split(X, y, labels):
+            yield train, test
+
+    def n_splits(self, X=None, y=None, labels=None):
+        """Returns the number of splitting iterations in the cross-validator
+
+        Parameters
+        ----------
+        X : (Ignored, exists for compatibility.)
+        y : (Ignored, exists for compatibility.)
+        labels : (Ignored, exists for compatibility.)
+        """
+        return self.n_folds
+
+
+class KFold(_BaseKFold):
+    """K-Folds cross-validator
+
+    Provides train/test indices to split data in train/test sets. Split
+    dataset into k consecutive folds (without shuffling).
+
+    Each fold is then used a validation set once while the k - 1 remaining
+    fold form the training set.
+
+    Read more in the :ref:`User Guide <cross_validation>`.
+
+    Parameters
+    ----------
+    n_folds : int, default=3
+        Number of folds. Must be at least 2.
+
+    shuffle : boolean, optional
+        Whether to shuffle the data before splitting into batches.
+
+    random_state : None, int or RandomState
+        Pseudo-random number generator state used for random
+        sampling. If None, use default numpy RNG for shuffling
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import KFold
+    >>> X = np.array([[1, 2], [3, 4], [1, 2], [3, 4]])
+    >>> y = np.array([1, 2, 3, 4])
+    >>> kf = KFold(n_folds=2)
+    >>> kf.n_splits(X)
+    2
+    >>> print(kf)  # doctest: +NORMALIZE_WHITESPACE
+    KFold(n_folds=2, random_state=None, shuffle=False)
+    >>> for train_index, test_index in kf.split(X):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [2 3] TEST: [0 1]
+    TRAIN: [0 1] TEST: [2 3]
+
+    Notes
+    -----
+    The first ``n % n_folds`` folds have size ``n // n_folds + 1``,
+    other folds have size ``n // n_folds``, where ``n`` is the number of
+    samples.
+
+    See also
+    --------
+    StratifiedKFold
+        For taking label information into account to avoid building folds with
+        imbalanced class distributions (for binary or multiclass
+        classification tasks).
+    """
+
+    def __init__(self, n_folds=3, shuffle=False,
+                 random_state=None):
+        super(KFold, self).__init__(n_folds, shuffle, random_state)
+        if shuffle:
+            self._rng = check_random_state(self.random_state)
+        self._shuffle = shuffle
+
+    def _iter_test_indices(self, X, y=None, labels=None):
+        n = _num_samples(X)
+        idxs = np.arange(n)
+        if self._shuffle:
```

Comment:
This is a very subtle behavior change! I'm not sure that is a great idea. What was the motivation for this?